### PR TITLE
Core: Allow building unsorted orders

### PR DESCRIPTION
**Pull request by aokolnychyi**
This PR allows building empty sort orders. Before that, it was failing with a validation exception.

### CODE REVIEW

**Comment by aokolnychyi** [comment]
cc @rdblue @RussellSpitzer

File: api/src/main/java/org/apache/iceberg/SortOrder.java [hunk 1]
```diff
@@ -258,14 +257,20 @@ Builder addSortField(String transformAsString, int sourceId, SortDirection direc
...
+      if (orderId != null && orderId == 0 && fields.size() != 0) {
         throw new IllegalArgumentException("Sort order ID 0 is reserved for unsorted order");
       }
-      if (fields.size() == 0 && orderId != 0) {
+      if (fields.size() == 0 && orderId != null && orderId != 0) {
         throw new IllegalArgumentException("Unsorted order ID must be 0");
       }
 
-      SortOrder sortOrder = new SortOrder(schema, orderId, fields);
+      if (fields.isEmpty()) {
```
Comment (RussellSpitzer) [inline api/src/main/java/org/apache/iceberg/SortOrder.java:267]: Could we do this first and then skip all the orderId / size checks? We could probably set order ID back to 1 as well?

Comment (RussellSpitzer) [reply api/src/main/java/org/apache/iceberg/SortOrder.java:267]: maybe

if (fields empty)
  if (orderId == 0)
      return empty sort order
  else
       Illegal orderId for empty sort order

if (orderId == 0)
    Illegal orderID for nonEmpty sort order

Comment (aokolnychyi) [reply api/src/main/java/org/apache/iceberg/SortOrder.java:267]: I have two options here:

```
if (fields.isEmpty()) {
  if (orderId == null || orderId == 0) {
    return SortOrder.unsorted();
  } else {
    throw new IllegalArgumentException("Unsorted order ID must be 0");
  }
} else if (orderId != null && orderId == 0) {
  throw new IllegalArgumentException("Sort order ID 0 is reserved for unsorted order");
}
```

or

```
if (fields.isEmpty()) {
  if (orderId != null && orderId != 0) {
    throw new IllegalArgumentException("Unsorted order ID must be 0");
  }
  return SortOrder.unsorted();
}

if (orderId != null && orderId == 0) {
  throw new IllegalArgumentException("Sort order ID 0 is reserved for unsorted order");
}
```

I tend to like the second option a bit more. What do you think, @RussellSpitzer?

Comment (RussellSpitzer) [reply api/src/main/java/org/apache/iceberg/SortOrder.java:267]: I think either is good, really I just wanted to have the control flow sorted like you have it now where we can clearly see the branch for unsorted vs sorted code, so beyond that I think it's your preference

Comment (aokolnychyi) [reply api/src/main/java/org/apache/iceberg/SortOrder.java:267]: Agree.

Comment (aokolnychyi) [reply api/src/main/java/org/apache/iceberg/SortOrder.java:260]: Updated. Resolving this.

File: api/src/main/java/org/apache/iceberg/SortOrder.java [hunk 2]
```diff
@@ -258,14 +257,20 @@ Builder addSortField(String transformAsString, int sourceId, SortDirection direc
...
+        if (orderId != null && orderId != 0) {
+          throw new IllegalArgumentException("Unsorted order ID must be 0");
+        }
+        return SortOrder.unsorted();
       }
-      if (fields.size() == 0 && orderId != 0) {
-        throw new IllegalArgumentException("Unsorted order ID must be 0");
+
+      if (orderId != null && orderId == 0) {
```
Comment (rdblue) [inline api/src/main/java/org/apache/iceberg/SortOrder.java:267]: If `orderId == 0`, then it can't be null, so the null check is redundant.

**Review by rdblue — APPROVED** [review]

**Comment by rdblue** [comment]
Looks good, one minor nit. Up to you whether to fix or commit.

File: api/src/main/java/org/apache/iceberg/SortOrder.java (hunk 2 above)
Comment (aokolnychyi) [reply api/src/main/java/org/apache/iceberg/SortOrder.java:267]: After this change, `orderId` is no longer a primitive. This check avoids an NPE.

**Comment by aokolnychyi** [comment]
@rdblue, it looks to me that the check is needed. Let me know what you think. I'll submit a follow-up if needed.

**Pull request merged**